devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,189 @@
+
+    def __getattr__(self, name):
+        if name == 'fd' or name == '_fd':
+            raise AttributeError()
+        if name in self.fd:
+            # self.fd[name] = (property, description), calls self.fd[name][0]
+            return self.fd[name][0](self)
+        return self.__getattribute__(name)
+
+    def xreplace(self, rule):
+        if self.is_Function:
+            return super(Differentiable, self).xreplace(rule)
+        else:
+            return self._expr.xreplace(rule)
+
+    def _merge_fd_properties(self, other):
+        merged = {}
+        merged["space_order"] = np.min([getattr(self, 'space_order', 100) or 100,
+                                        getattr(other, 'space_order', 100)])
+        merged["time_order"] = np.min([getattr(self, 'time_order', 100) or 100,
+                                       getattr(other, 'time_order', 100)])
+        merged["indices"] = tuple(filter_ordered(self.indices +
+                                                 getattr(other, 'indices', ())))
+        merged["fd"] = dict(getattr(self, 'fd', {}), **getattr(other, 'fd', {}))
+        merged["staggered"] = self.staggered
+        merged["dtype"] = self.dtype
+        return merged
+
+    def __add__(self, other):
+        return Differentiable(sympy.Add(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    __iadd__ = __add__
+    __radd__ = __add__
+
+    def __sub__(self, other):
+        return Differentiable(sympy.Add(*[getattr(self, '_expr', self),
+                                          -getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    def __rsub__(self, other):
+        return Differentiable(sympy.Add(*[-getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    __isub__ = __sub__
+
+    def __mul__(self, other):
+        return Differentiable(sympy.Mul(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    __imul__ = __mul__
+    __rmul__ = __mul__
+
+    def __truediv__(self, other):
+        return Differentiable(sympy.Mul(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)**(-1)]),
+                              **self._merge_fd_properties(other))
+
+    def __rtruediv__(self, other):
+        return Differentiable(sympy.Mul(*[getattr(other, '_expr', other),
+                                          getattr(self, '_expr', self)**(-1)]),
+                              **self._merge_fd_properties(other))
+
+    __floordiv__ = __truediv__
+    __rdiv__ = __rtruediv__
+    __div__ = __truediv__
+    __rfloordiv__ = __rtruediv__
```

Comment:
Note sure where to leave this comment... anyway...
we're missing __mod__, __int__, and __float__ at least